src/features/wallet/root/ActivityItem.tsx

```diff
@@ -1,49 +1,175 @@
+
+  const amount = useMemo(() => {
+    if (item instanceof AddGatewayV1) {
+      return animalHash(item.gateway)
+    }
+    if (item instanceof AssertLocationV1 || item instanceof TransferHotspotV1) {
+      return `-${item.fee?.toLocaleString()}`
+    }
+    if (item instanceof RewardsV1 || item instanceof PaymentV2) {
+      return `+${item.totalAmount.toString()}`
+    }
+    if (item instanceof PaymentV1) {
+      return `+${item.amount.floatBalance.toString()}`
+    }
+    const pendingTxn = item as PendingTransaction
+    if (pendingTxn.txn !== undefined) {
+      if (pendingTxn.txn.type === 'add_gateway_v1') {
+        return animalHash((pendingTxn.txn as AddGatewayV1).gateway)
+      }
+      return `-${pendingTxn.txn.fee.toLocaleString()}`
+    }
 
-  const theme = useTheme<Theme>()
+    return ''
+  }, [item])
 
-  const colors: Record<string, string> = {
-    rewards: theme.colors.purpleBright,
-    sent: theme.colors.blueBright,
-    received: theme.colors.greenMain,
-    add: theme.colors.purple100,
+  const time = useMemo(() => {
+    let val: Date
+    const pending = item as PendingTransaction
+    if (pending.txn !== undefined) {
+      val = new Date(pending.createdAt)
+    } else {
+      val = fromUnixTime((item as AddGatewayV1).time)
+    }
+
+    if (val) {
+      return formatDistanceToNow(val, {
+        locale: shortLocale,
+        addSuffix: true,
+      })
+    }
+
+    return undefined
+  }, [item])
+
+  const titles = useCallback(
+    (type: string) => {
+      if (!TxnTypeKeys.find((k) => k === type)) {
+        return type
+      }
+
+      switch (type as TxnType) {
+        case 'add_gateway_v1':
+          return t('transactions.added')
+        case 'payment_v1':
+        case 'payment_v2':
+          return isSending ? t('transactions.sent') : t('transactions.received')
+        case 'assert_location_v1':
+          return t('transactions.location')
+        case 'transfer_hotspot_v1':
+          return t('transactions.transfer')
+        case 'rewards_v1':
+          return t('transactions.mining')
+      }
+    },
+    [isSending, t],
+  )
+
+  const iconBackgroundColor = useCallback(
+    (type: string) => {
+      if (!TxnTypeKeys.find((k) => k === type)) {
+        return colors.primaryBackground
+      }
+
+      switch (type as TxnType) {
+        case 'transfer_hotspot_v1':
+        case 'add_gateway_v1':
+          return colors.purple100
+        case 'payment_v1':
+        case 'payment_v2':
+          return isSending ? colors.blueBright : colors.greenMain
+        case 'assert_location_v1':
+          return colors.purpleMuted
+        case 'rewards_v1':
+          return colors.purpleBright
+      }
+    },
+    [isSending, colors],
+  )
+
+  const icon = useCallback(
+    (type: string) => {
+      if (!TxnTypeKeys.find((k) => k === type)) {
+        return null
+      }
+
+      const size = 24
+      switch (type as TxnType) {
+        case 'transfer_hotspot_v1':
```

Comment:
Should we save these types in an enum or something somewhere? Or do they type match already?

Reply:
TxnType gives us type safety here.

Reply:
Cool, works for me